zerver/lib/thumbnail.py

```diff
@@ -39,122 +40,106 @@ def generate_thumbnail_url(path: str, size: str = "0x0") -> str:
+
     # This function returns three values:
     # 1) Emoji image data.
-    # 2) If emoji is gif i.e. animated.
-    # 3) If is animated then return still image data i.e. first frame of gif.
-
-    try:
-        im = Image.open(io.BytesIO(image_data))
-        image_format = im.format
-        if getattr(im, "n_frames", 1) > 1:
-            # There are a number of bugs in Pillow which cause results
-            # in resized images being broken. To work around this we
-            # only resize under certain conditions to minimize the
-            # chance of creating ugly images.
-            should_resize = (
-                im.size[0] != im.size[1]  # not square
-                or im.size[0] > MAX_EMOJI_GIF_SIZE  # dimensions too large
-                or len(image_data) > MAX_EMOJI_GIF_FILE_SIZE_BYTES  # filesize too large
+    # 2) If the emoji is animated.
+    # 3) If it is animated, the still image data i.e. first frame of gif.
+    with libvips_check_image(image_data) as source_image:
+        if source_image.get_n_pages() == 1:
+            return (
+                pyvips.Image.thumbnail_buffer(
+                    image_data,
+                    size,
+                    height=size,
+                    crop=pyvips.Interesting.CENTRE,
+                ).write_to_buffer(write_file_ext),
+                False,
+                None,
             )
-
-            # Generate a still image from the first frame.  Since
-            # we're converting the format to PNG anyway, we resize unconditionally.
-            still_image = im.copy()
-            still_image.seek(0)
-            still_image = ImageOps.exif_transpose(still_image)
-            still_image = ImageOps.fit(still_image, (size, size), Image.Resampling.LANCZOS)
-            out = io.BytesIO()
-            still_image.save(out, format="PNG")
-            still_image_data = out.getvalue()
-
-            if should_resize:
-                image_data = resize_animated(im, size)
-
-            return image_data, True, still_image_data
-        else:
-            # Note that this is essentially duplicated in the
-            # still_image code path, above.
-            im = ImageOps.exif_transpose(im)
-            im = ImageOps.fit(im, (size, size), Image.Resampling.LANCZOS)
-            out = io.BytesIO()
-            im.save(out, format=image_format)
-            return out.getvalue(), False, None
-    except OSError:
-        raise BadImageError(_("Could not decode image; did you upload an image file?"))
-    except DecompressionBombError:
-        raise BadImageError(_("Image size exceeds limit."))
+        first_still = pyvips.Image.thumbnail_buffer(
+            image_data,
+            size,
+            height=size,
+            crop=pyvips.Interesting.CENTRE,
+        ).write_to_buffer(".png")
+
+        animated = pyvips.Image.thumbnail_buffer(
+            image_data,
+            size,
+            height=size,
+            option_string="n=-1",
```

Comment:
Maybe worth a comment describing or linking to what this `option_string` does? I was having trouble finding it in upstream's docs immediately.